### ISSUE TITLE
log_prob() for tf.distributions.Dirichlet returns values > 0

### ISSUE THREAD

**Issue by josago** [tensorflow/tensorflow#20587]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution**: Debian Stretch, 64 bits
- **TensorFlow installed from**: virtualenv / pip
- **TensorFlow version**: 1.9.0rc2
- **Python version**: 3.5
- **Bazel version**: N/A
- **GCC/Compiler version (if compiling from source)**: N/A
- **CUDA/cuDNN version**: 9.0/7.0
- **GPU model and memory**: Geforce GTX 1080 Ti, 11172 MiB
- **Exact command to reproduce**: See the description and source code below.

### Describe the problem
The method log_prob() for tf.distributions.Dirichlet sometimes returns values greater than 0, which is not mathematically possible.

### Source code / logs
Piece of code that fails:

```
dist = tf.distributions.Dirichlet(
    tf.exp(log_alphas),
    validate_args=True,
    allow_nan_stats=False,
)

objective = tf.reduce_sum(dist.log_prob(target))
```

Manual solution that works properly, using the mathematical definition for the Dirichlet:

```
def dirichlet_log_prob(log_alphas, x):
    return tf.reduce_sum(
        tf.lgamma(tf.reduce_logsumexp(log_alphas, axis=1)) +
        tf.reduce_sum(tf.multiply(tf.exp(log_alphas) - 1, tf.log(x)) - tf.lgamma(log_alphas), axis=1)
    )

objective = dirichlet_log_prob(log_alphas, target)
```

Data used in both examples:

```
log_alphas = [
    [-1.6231717, -2.112546 , -2.1379914, -2.0885904, -1.9052815, -1.759051 ],
    [-1.4579483, -1.8557736, -1.9981235, -1.8574138, -1.5318792, -1.3888079],
    [-1.4608178, -1.8787189, -1.9842598, -1.8636204, -1.6005809, -1.4765729],
    [-1.4781257, -1.8943347, -2.0051253, -1.880945 , -1.613416 , -1.4802316],
    [-1.447619 , -1.9153082, -1.9177424, -1.8622615, -1.7686105, -1.712473 ],
    [-1.490995 , -2.0010724, -2.0137913, -1.9128757, -1.8797761, -1.8842031],
    [-1.5973192, -2.0856535, -2.1033392, -2.0476089, -1.9143845, -1.8061923],
    [-1.5467792, -2.0158603, -2.0372252, -1.9859502, -1.8327508, -1.7130013]
]
target = [
    [2.7760255e-01, 3.2513769e-06, 3.6051267e-01, 3.2513769e-06, 8.0276497e-02, 2.8160176e-01],
    [6.5851367e-01, 3.4333350e-06, 1.7245641e-01, 3.4333350e-06, 2.1149343e-02, 1.4787373e-01],
    [3.1866562e-02, 2.0533266e-02, 3.3333222e-06, 2.0566598e-02, 6.1813128e-01, 3.0889896e-01],
    [3.4036295e-06, 1.3675784e-01, 3.4036295e-06, 8.6322856e-01, 3.4036295e-06, 3.4036295e-06],
    [1.6921267e-01, 2.1246435e-02, 7.6704454e-01, 2.1246435e-02, 2.1246435e-02, 3.4490965e-06],
    [3.4373238e-06, 4.7641307e-02, 3.4373238e-06, 9.5234495e-01, 3.4373238e-06, 3.4373238e-06],
    [1.3007499e-01, 4.5678858e-02, 6.3374266e-02, 7.3981225e-01, 3.4293435e-06, 2.1056170e-02],
    [5.3522962e-01, 3.5806104e-06, 1.7695376e-01, 3.5806104e-06, 4.4113118e-02, 2.4369633e-01]
]
```

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Bazel version

**Comment by josago** [comment]
Updated the info now; I'm using virtualenv with PyCharm. No bazel or gcc within the venv/ folder; the GCC version on the host is 6.3.0.

**Comment by josago** [comment]
log_prob() can indeed be >0 for some configurations of alpha.

**Comment by srvasude** [comment]
Hi!
  As mentioned, note that the log_prob() > 0 for small alphas. Take for example a beta distribution. When the concentration parameters are small, the mass is very much concentrated around zero and 1, and unlikely in the middle. This causes the pdf function to be very large at small and large values, causing the log_pdf to be > 0.

Note that in the implementation above, the normalizing constant is computed by using lgamma(log_alpha) instead of lgamma(alpha) which causes the discrepancy.
